be/src/runtime/runtime_query_statistics_mgr.cpp

```diff
@@ -17,14 +17,432 @@
+            auto msg = fmt::format(
+                    "Register fragment profile {} {} failed, load channel profile is null",
+                    print_id(query_id), -1);
+            DCHECK(false) << msg;
+            LOG_ERROR(msg);
+            continue;
+        }
+
+        load_channel_profiles_req.push_back(*load_channel_profile);
+    }
+
+    if (load_channel_profiles_req.size() > 0) {
+        profile.__set_load_channel_profiles(load_channel_profiles_req);
+    }
+
+    TReportExecStatusParams req;
+    req.__set_query_profile(profile);
+    req.__set_query_id(TUniqueId());
+
+    return req;
+}
+
+TReportExecStatusParams RuntimeQueryStatiticsMgr::create_report_exec_status_params_non_pipeline(
+        const TUniqueId& query_id,
+        const std::unordered_map<TUniqueId, std::shared_ptr<TRuntimeProfileTree>>&
+                instance_id_to_profile,
+        const std::vector<std::shared_ptr<TRuntimeProfileTree>>& load_channel_profile) {
+    TQueryProfile profile;
+    std::vector<TUniqueId> fragment_instance_ids;
+    std::vector<TRuntimeProfileTree> instance_profiles;
+    std::vector<TRuntimeProfileTree> load_channel_profiles;
+
+    for (auto entry : instance_id_to_profile) {
+        TUniqueId instance_id = entry.first;
+        std::shared_ptr<TRuntimeProfileTree> profile = entry.second;
+
+        if (profile == nullptr) {
+            auto msg = fmt::format("Register instance profile {} {} failed, profile is null",
+                                   print_id(query_id), print_id(instance_id));
+            DCHECK(false) << msg;
+            LOG_ERROR(msg);
+            continue;
+        }
+
+        fragment_instance_ids.push_back(instance_id);
+        instance_profiles.push_back(*profile);
+    }
+
+    profile.__set_query_id(query_id);
+    profile.__set_fragment_instance_ids(fragment_instance_ids);
+    profile.__set_instance_profiles(instance_profiles);
+    profile.__set_load_channel_profiles(load_channel_profiles);
+
+    TReportExecStatusParams res;
+    res.__set_query_profile(profile);
+    res.__set_query_id(TUniqueId());
+    return res;
+}
+
+void RuntimeQueryStatiticsMgr::start_report_thread() {
+    if (started.load()) {
+        DCHECK(false) << "report thread has been started";
+        LOG_ERROR("report thread has been started");
+        return;
+    }
+
+    started.store(true);
+
+    for (size_t i = 0; i < config::report_exec_status_thread_num; ++i) {
+        this->_report_profile_threads.emplace_back(std::make_unique<std::thread>(
+                &RuntimeQueryStatiticsMgr::report_query_profiles_thread, this));
+    }
+}
+
+void RuntimeQueryStatiticsMgr::report_query_profiles_thread() {
+    while (true) {
+        {
+            std::unique_lock<std::mutex> lock(_report_profile_mutex);
+
+            while (_query_profile_map.empty() && _profile_map_x.empty() &&
+                   !_report_profile_thread_stop) {
+                _report_profile_cv.wait(lock);
```

Comment:
wait 3s and check again, or will hang

Reply:
In what situation would this thread hang forever ?

Reply:
done